### PR TITLE
feat(mcp): add namespace tracking for mcp tools

### PR DESCRIPTION
**Pull request by dhable**
#### feat(mcp): add namespace tracking for mcp tools

Track the mcp server key as a namespace for each tool to provide additional context in governance, observability, and approval flows. The namespace is captured from the config during tool registration and propagated through tool calls.

The namespace appears in:
- otel `mcp.tool_call` spans as an attribute
- governance catalog entries for policy evaluation
- hitl approval requests and matching patterns

Hitl patterns support namespace-scoped matching via `<namespace>:<name>` syntax, where both segments accept glob patterns. The first `:` serves as the delimiter.

The namespace is not exposed to the model since OpenAI and Anthropic do not support SEP-986, making it difficult to choose a namespace/name delimiter that works across providers.

Fixes: #712 

---

#### fix(mcp): warn if multiple servers use same tool name

It's possible for two different MCP servers to expose a tool
using the same name. Aura now checks for this on boot and
on governance sync to warn the user about their configuration.

Fixes: https://github.com/mezmo/aura/issues/713

### CODE REVIEW

File: crates/aura/src/mcp/types.rs [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        &self.0
+    }
+
+    pub fn matches(&self, pattern: &str) -> bool {
+        let part: Vec<&str> = pattern.split(NAMESPACE_DELIMETER).collect();
+        let tool_pattern = if part.len() > 1 {
+            &part[1..].join("")
+        } else {
+            pattern
+        };
```
[the remark this answers is not in the document]
Comment (dhable) [reply crates/aura/src/mcp/types.rs]: A colon is not a valid pattern according to SEP-986 so it would never appear in the tool name. Since namespaces should also follow the SEP-986 pattern, a colon is not valid in the namespace. Dropping them would happen anyway after sanitization of the tool name so this isn't an issue.

File: crates/aura/src/hitl/gate.rs [hunk 2]
```diff
@@ -114,13 +115,32 @@ impl HitlApprovalWrapper {
     /// approval tool itself ("request_approval" == RequestApprovalTool::NAME).
     /// Any match gates the call; the returned pattern is only the reported
     /// `origin.matched_pattern`, so pattern order has no effect on gating.
-    fn matched_pattern(&self, tool_name: &str) -> Option<&str> {
+    ///
+    /// Namespace-aware: a pattern containing `/` (e.g. `github/*`) scopes the
+    /// match to `tool_namespace` — see [`crate::mcp::tool_pattern_matches`].
+    /// `tool_namespace` is `None` for tools with no known MCP server (e.g.
+    /// filesystem or client-side tools), which only patterns without `/` can
+    /// match.
```
[the remark this answers is not in the document]
Comment (dhable) [reply crates/aura/src/hitl/gate.rs]: Is this still a concern?

File: crates/aura/src/tool_wrapper.rs [hunk 3]
```diff
@@ -71,6 +71,12 @@ use crate::mcp::CallOutcome;
 pub struct ToolCallContext {
     /// Tool name being called
     pub tool_name: String,
+    /// The MCP server (config key) `tool_name` came from, when known. Set by
+    /// `Agent::add_mcp_tool` for MCP-originated tools so wrapper-layer
+    /// consumers (HITL approval requests, in particular) can attribute the
+    /// call to its server without it ever being folded into `tool_name`,
+    /// which is always the bare name sent to the model.
+    pub tool_namespace: Option<String>,
```
[the remark this answers is not in the document]
Comment (dhable) [reply crates/aura/src/tool_wrapper.rs]: Is this still a concern?

File: crates/aura/src/hitl/gate.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhable) [reply crates/aura/src/hitl/gate.rs]: Are there other stale comments?

[the remark this answers is not in the document]
Comment (dhable) [reply crates/aura/src/hitl/gate.rs]: Any more?

[the remark this answers is not in the document]
Comment (dhable) [reply crates/aura/src/hitl/gate.rs]: Is there anything else?

[the remark this answers is not in the document]
Comment (dhable) [reply crates/aura/src/hitl/gate.rs]: This is fine for now.

File: crates/aura/src/orchestration/overview.rs [hunk 4]
```diff
@@ -112,43 +110,6 @@ async fn discover_tools(
...
-///
-/// These are the values AURA holds, not the ones the server advertised:
-/// `McpManager::sanitize_mcp_tool` has already rewritten the name to the
-/// LLM-safe character set and, under `[mcp].sanitize_schemas`, rewritten the
-/// input schema. Publishing that form is what makes the output usable for
-/// governance — it names the tools AURA actually invokes with the schemas the
-/// model actually receives.
-///
-/// `icons` is dropped; everything else in the MCP `Tool` object carries over.
-fn tool_overview(tool: &rmcp::model::Tool) -> McpToolOverview {
```
Comment (dhable) [inline crates/aura/src/orchestration/overview.rs:125]: This was turned into an implementation of `Into` on the tooling type itself. It did so much reaching into the raw struct that it seemed like closer to the type definition was the correct fit.

https://github.com/mezmo/aura/pull/700/changes#diff-1cfaaa68f58e8e0b112f2493d9c2ec708604575b3b6ad05f84ddd802fde6eae5R156-R191

File: crates/aura/src/hitl/gate.rs [hunk 5]
```diff
@@ -114,13 +115,28 @@ impl HitlApprovalWrapper {
...
             return None;
         }
+
+        let tool_name = ToolName::from(tool_name);
+        let tool_ns = tool_namespace.map(ToolNamespace::from);
         self.patterns
             .iter()
-            .find(|p| p.matches(tool_name))
+            .find(|p| {
+                let mut is_match = tool_name.matches(p.as_str());
```
Comment (jakedipity) [inline crates/aura/src/hitl/gate.rs:133]: `ToolName::matches` uses a hand rolled `glob_match` function that doesn't satisfy all the expectations (e.g. alternation and character classes) like `GlobPattern::matches` does, so this is a clear regression.

I imagine we don't want to be using an incomplete implementation for glob matching and can use two `GlobPattern`'s and splitting by colon manually.

Additionally dug up an even worse hand rolled implementation in crates/aura-cli/src/permissions.rs:584 so this might warrant more cleanup in a follow up PR.

**Review by jakedipity — CHANGES_REQUESTED** [review]

File: crates/aura/src/hitl/gate.rs (hunk 5 above)
Comment (dhable) [reply crates/aura/src/hitl/gate.rs]: I opened issue #705 to handle the further cleanup on glob matching. This went through a bit of rework. Now the config uses the `GlobPattern`type in the MCP filter fields that I could identify. It now understands the namespace pattern and splits into two.

There's more refactoring that could be done to harden up the types for tool names/namespaces but that was quickly escalating. I decided to cap it here and revisit it more in the future.

**Review by Shearerbeard — DISMISSED** [review]

**Comment by teriyakichild** [comment]
Pulled this down and read through it. A few things I'd like to see before merge.

**Reject duplicate bare names in `add_all_tools`, or change `Fixes: #186` to `Ref:`**
#186 is about silent overwrite when several servers expose the same tool name. This PR tracks the namespace as metadata but leaves registration as-is: Rig's `ToolSet` is a `HashMap` keyed by name and aura registers from a `HashMap` too, so the last insert wins and the winner can change per restart. Before this PR that was a routing hazard. With `prod:*` HITL patterns it becomes a gate that fires only when prod happened to win. Since the loop already filters per server, tracking seen bare names and failing the build with both server names on a repeat would close it. Configs that scope each agent to one server with `mcp_filter` are unaffected. If that's out of scope here, the footer should reference #186 rather than close it.

**Validate pattern segments in `GlobPattern::new`** (`crates/aura-config/src/config.rs:1765`)
`split_once` accepts `k8s:`, `:get_pods`, `k8s: *` and `k8s:pods:*` without error. globset compiles an empty segment as a glob that only matches the empty string, so all of these load fine and never match anything. In `mcp_filter` that just drops a tool. In `require_approval` it means the tool runs ungated, and nothing logs why. The SEP-986 argument on the earlier thread covers tool and namespace *data*, not what a user types into a pattern. Rejecting an empty segment or a second colon at load turns a typo into a config error.

**One end-to-end test for scoped HITL gating**
The stamp in `add_mcp_tool` is the only place a namespace reaches the approval gate. `matched_pattern_is_namespace_aware` tests the matcher in isolation, but nothing builds an agent with a `github:*` pattern and asserts the gate sees `tool_namespace = Some("github")`. The three-transport manager fixture in `builder.rs` tests looks like the right home.

**Small ones**
- `McpProbeDispatcher` finds the server by name and then passes `None` for the namespace to `execute_mcp_tool`. Returning the key from that lookup gives probe spans a real value.
- `ATTR_TOOL_NAMESPACE` in `logging.rs` is unreferenced; the span uses the literal. Also a missing namespace records as `""` rather than being omitted.
- `From<&str> for GlobPattern` is a public conversion that panics on a bad glob. `TryFrom`, or gating it to tests, keeps that out of library callers.
- The crate now has two public `ToolName`s: the `String` alias in `tool_event_broker.rs` and the new newtype in `mcp/types.rs`.

**Comment by dhable** [comment]
re:
> Validate pattern segments in GlobPattern::new (crates/aura-config/src/config.rs:1765)
split_once accepts k8s:, :get_pods, k8s: * and k8s:pods:* without error. globset compiles an empty segment as a glob that only matches the empty string, so all of these load fine and never match anything. In mcp_filter that just drops a tool. In require_approval it means the tool runs ungated, and nothing logs why. The SEP-986 argument on the earlier thread covers tool and namespace data, not what a user types into a pattern. Rejecting an empty segment or a second colon at load turns a typo into a config error.

It was too easy to continue coming up with edge cases that broke things when trying to only use a regex and the `globset` crate. Defining a simple set of rules in PEG plug all the holes in validation. Shipped that into the config along with a slew of test cases.

re:
> ATTR_TOOL_NAMESPACE in logging.rs is unreferenced; the span uses the literal. Also a missing namespace records as "" rather than being omitted.

That was left over from some earlier thoughts. Deleted it.

re:
> From<&str> for GlobPattern is a public conversion that panics on a bad glob. TryFrom, or gating it to tests, keeps that out of library callers.

We have a lot of automated tests that got even more verbose when they had to turn `str`s into the type wrapper representations. I think its worth the panic in a unit test if the test input value is wrong, which is why this isn't `TryFrom`. The rub is that this crosses crate bounds so simply limiting it to the test config isn't going to work. 

I'm thinking about introducing a feature to aura-config, `test_utils`, that can be used to expose some of these less production refined functions that make writing tests a lot simpler.

**Comment by dhable** [comment]
re:
> McpProbeDispatcher finds the server by name and then passes None for the namespace to execute_mcp_tool. Returning the key from that lookup gives probe spans a real value.

It seems odd that the namespace for the tool would be detached from the `McpClient`, which is effectively the container that the namespace names (i.e. all of the tools handled by a McpClient also relate to the same namespace metadata field). Based on our code, it seems easiest to include the namespace on the client and then it can be referenced inside `execute_mcp_tool` without huge threading efforts.

re:
> The crate now has two public ToolNames: the String alias in tool_event_broker.rs and the new newtype in mcp/types.rs.

Dropping the type alias in favor of the type wrapper. Type aliases are still leaky about the underlying type and don't allow for implementing narrow interfaces. This aligns with how @Shearerbeard was also thinking about modeling domain concepts in the code base.

**Comment by dhable** [comment]
> ### Comments Outside Diff
> 
> These findings sit on lines the diff does not cover, so they could not be posted inline. Each one leaves this list once its file changes.
> 
>     * P1 **Worker completions lose attribution** `crates/aura/src/orchestration/orchestrator.rs:301` [▶](https://github.com/mezmo/aura/blob/7a7ae0b41ed9ae8a3441c94b6330cbca7e6172cb/crates/aura/src/orchestration/orchestrator.rs#L295-L301)
>       Observed MCP tool starts preserve `tool_initiator_id` through `by_worker`, but the corresponding completions carry no initiator identity and are always wrapped with `by_coordinator`. Because the event envelope's agent ID becomes the emitted `worker_id`, a worker tool call starts under the worker and completes under `coordinator`. Clients therefore associate the result with the wrong agent and cannot reliably correlate the tool lifecycle.

@greptile-app This is part of a larger change that is coming - https://github.com/mezmo/aura/pull/632#issuecomment-5722966305. Do not flag it in this PR anymore.

**Comment by dhable** [comment]
re:
> One end-to-end test for scoped HITL gating
The stamp in add_mcp_tool is the only place a namespace reaches the approval gate. matched_pattern_is_namespace_aware tests the matcher in isolation, but nothing builds an agent with a github:* pattern and asserts the gate sees tool_namespace = Some("github"). The three-transport manager fixture in builder.rs tests looks like the right home.

Generated some additional tests located at https://github.com/mezmo/aura/pull/700/changes#diff-69ced7c0f5daba4f46388a12e5aee66be7cca46a222e81967352591454ea4989R2182-R2366.

File: crates/aura/src/mcp/manager.rs [hunk 6]
```diff
@@ -914,27 +913,66 @@ impl McpManager {
...
+    /// One human-readable line per colliding tool name, naming every server
+    /// that claims it and the one that wins.
+    pub fn collision_report(&self) -> Vec<String> {
+        self.colliding_tool_names()
+            .into_iter()
+            .map(|(tool, servers)| {
+                let winner = servers.first().map(String::as_str).unwrap_or("");
+                format!(
+                    "MCP tool '{}' is advertised by {} servers ({}); only '{}' will be reachable. \
+                     Scope each agent with [agent].mcp_filter, or rename the tool on all but one \
```
[the remark this answers is not in the document]
Comment (dhable) [reply crates/aura/src/mcp/manager.rs:964]: I think this is fine for now. This warning isn't meant as a definitive evaluation, rather to alert the user that tool collisions might happen so they can either explore other MCP servers or prefixes on the MCP servers themselves to alleviate the problem.

This warning will be removed when issue #714 lands.

**Review by jakedipity — APPROVED** [review]
